File: Go-Solutions/nowcoder/nc43.go
package nowcoder

// https://leetcode-cn.com/problems/permutations/solution/hui-su-suan-fa-python-dai-ma-java-dai-ma-by-liweiw/
func permute43(num []int) [][]int {
	// write code here
	if len(num) == 0 {
		return [][]int{}
	}

	res := make([][]int, 0)
	used := make([]bool, len(num))
	list := make([]int, 0)

	dfs43(num, &res, &list, used)
	return res
}

func dfs43(num []int, res *[][]int, list *[]int, used []bool) {
	if len(num) == len(*list) {
		listToAdd := make([]int, 0)
		for i := 0; i < len(*list); i++ {
			listToAdd = append(listToAdd, (*list)[i])
		}
		*res = append(*res, listToAdd)
		return
	}

	for i := 0; i < len(num); i++ {
		if !used[i] {
			*list = append(*list, num[i])
			used[i] = true

			dfs43(num, res, list, used)

			*list = (*list)[:len(*list)-1]
			used[i] = false
		}
	}
}

File: Java-Kotlin/src/main/java/nowcoder/nc43.java
package nowcoder;

import java.util.ArrayList;

public class nc43 {

    private static void dfs(int[] num,
                            ArrayList<ArrayList<Integer>> res,
                            ArrayList<Integer> list,
                            boolean[] used) {
        if (num.length == list.size()) {
            res.add(new ArrayList<>(list));
            return;
        }

        for (int i = 0; i < num.length; i++) {
            if (!used[i]) {
                list.add(num[i]);
                used[i] = true;

                dfs(num, res, list, used);

                list.remove(list.size() - 1);
                used[i] = false;
            }
        }
    }

    // https://leetcode-cn.com/problems/permutations/solution/hui-su-suan-fa-python-dai-ma-java-dai-ma-by-liweiw/
    public ArrayList<ArrayList<Integer>> permute(int[] num) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        if (num == null || num.length == 0) {
            return res;
        }
        boolean[] used = new boolean[num.length];
        ArrayList<Integer> list = new ArrayList<>();
        dfs(num, res, list, used);
        return res;
    }
}
